cdap-api/src/main/java/co/cask/cdap/api/dataset/lib/CompositeDatasetDefinition.java

```diff
@@ -72,27 +115,55 @@ protected CompositeDatasetDefinition(String name, Map<String, ? extends DatasetD
   }
 
   @Override
-  public final DatasetSpecification configure(String instanceName, DatasetProperties properties) {
+  public DatasetSpecification configure(String instanceName, DatasetProperties properties) {
     List<DatasetSpecification> specs = new ArrayList<>();
     for (Map.Entry<String, ? extends DatasetDefinition> impl : this.delegates.entrySet()) {
       specs.add(impl.getValue().configure(impl.getKey(), properties));
     }
-
     return DatasetSpecification.builder(instanceName, getName())
       .properties(properties.getProperties())
       .datasets(specs)
       .build();
   }
 
   @Override
-  public final DatasetAdmin getAdmin(DatasetContext datasetContext, DatasetSpecification spec,
-                                     ClassLoader classLoader) throws IOException {
-    List<DatasetAdmin> admins = new ArrayList<>();
+  public DatasetSpecification reconfigure(String instanceName,
+                                          DatasetProperties newProperties,
+                                          DatasetSpecification currentSpec) throws IncompatibleUpdateException {
+    List<DatasetSpecification> specs = new ArrayList<>();
     for (Map.Entry<String, ? extends DatasetDefinition> impl : this.delegates.entrySet()) {
-      admins.add(impl.getValue().getAdmin(datasetContext, spec.getSpecification(impl.getKey()), classLoader));
+      specs.add(reconfigure(impl.getValue(), impl.getKey(),
+                            newProperties, currentSpec.getSpecification(impl.getKey())));
     }
+    return DatasetSpecification.builder(instanceName, getName())
+      .properties(newProperties.getProperties())
+      .datasets(specs)
+      .build();
+  }
 
+  @Override
+  public final DatasetAdmin getAdmin(DatasetContext datasetContext, DatasetSpecification spec,
+                                     ClassLoader classLoader) throws IOException {
+    if (1 == delegates.size()) {
+      // for a single delegate, we don't need a composite admin
+      return getAdmin(datasetContext, delegates.keySet().iterator().next(), spec, classLoader);
+    }
+    Map<String, DatasetAdmin> admins = new HashMap<>();
+    for (String name : this.delegates.keySet()) {
+      admins.put(name, getAdmin(datasetContext, name, spec, classLoader));
+    }
     return new CompositeDatasetAdmin(admins);
   }
+
+  private DatasetAdmin getAdmin(DatasetContext datasetContext, String name,
+                                DatasetSpecification spec, ClassLoader classLoader) throws IOException {
+    return delegates.get(name).getAdmin(datasetContext, spec.getSpecification(name), classLoader);
+  }
+
+  protected <A extends Dataset, DS extends DatasetAdmin> DatasetDefinition<A, DS> getDelegate(String name) {
```

Comment:
Better use `DS` for `Dataset` and `A` for `DatasetAdmin` (it is reversed right now).